Lib/exception_group.py

```diff
@@ -0,0 +1,215 @@
+
+import sys
+
+
+class TracebackGroup:
+    def __init__(self, excs):
+        self.tb_next_map = {} # exception id to tb
+        for e in excs:
+            if isinstance(e, ExceptionGroup):
+                for e_ in e.excs:
+                    self.tb_next_map[id(e_)] = e_.__traceback__
+            else:
+                if e.__traceback__:
+                    self.tb_next_map[id(e)] = e.__traceback__.tb_next
+                else:
+                    self.tb_next_map[id(e)] = None
+
+class ExceptionGroup(BaseException):
+
+    def __init__(self, excs, *, tb=None):
+        """ Construct a new ExceptionGroup
+
+        excs: sequence of exceptions
+        tb [optional]: the __traceback__ of this exception group.
+        Typically set when this ExceptionGroup is derived from another.
+        """
+        self.excs = excs
+        # self.__traceback__ is updated as usual, but self.__traceback_group__
+        # is set when the exception group is created.
+        # __traceback_group__ and __traceback__ combine to give the full path.
+        self.__traceback__ = tb
+        self.__traceback_group__ = TracebackGroup(self.excs)
+
+    def project(self, condition):
+        """ Split an ExceptionGroup based on an exception predicate
+
+        returns two new ExceptionGroups: match, rest of the exceptions
+        of self for which condition(e) returns True and False, respectively.
+        match and rest have the same nested structure as self, but empty
+        sub-exceptions are not included.
+
+        condition: BaseException --> Boolean
+        """
+        match, rest = [], []
+        for e in self.excs:
+            if isinstance(e, ExceptionGroup): # recurse
+                e_match, e_rest = e.project(condition)
+                if e_match:
+                    match.append(e_match)
+                if e_rest:
+                    rest.append(e_rest)
+            else:
+                if condition(e):
+                    match.append(e)
+                    e_match, e_rest = e, None
+                else:
+                    rest.append(e)
+        return (ExceptionGroup(match, tb=self.__traceback__),
+                ExceptionGroup(rest, tb=self.__traceback__))
+
+    def split(self, type):
+        """ Split an ExceptionGroup to extract exceptions of type E
+
+        type: An exception type
+        """
+        return self.project(lambda e: isinstance(e, type))
+
+    def subgroup(self, keep):
+        """ Split an ExceptionGroup to extract only exceptions in keep
+
+        keep: List[BaseException]
+        """
+        match, _ = self.project(lambda e: e in keep)
+        return match
+
+    def extract_traceback(self, exc):
+        """ returns the traceback of a single exception
+
+        If exc is in this exception group, return its
+        traceback as a list of frames. Otherwise, return None.
```

Comment:
Should probably also return the other info in the tb object (tb_lineno, tb_lasti). Possibly reuse one of the named tuple types from traceback.py?